File: Difficulty/Medium/[92] Reverse Linked List II/solution.cpp
/*
 * @lc app=leetcode id=92 lang=cpp
 *
 * [92] Reverse Linked List II
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <vector>
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        if (m == n) return head;
        m = min(m, n);
        n = max(m, n);
        vector<ListNode*> pre((n - m + 1) / 2, nullptr);
        ListNode* new_head = new ListNode(0);
        new_head->next = head;
        int count = 1;
        while (head != NULL) {
            if (count >= m && count < (n + m + 1) / 2) {
                pre[count - m] = head;
            }
            if (count > (m + n) / 2 && count <= n){
                swapListNodeVal(pre[n - count], head);
            }
            count++;
            head = head->next;
        }
        return new_head->next;
    }

    void swapListNodeVal(ListNode *A, ListNode *B) {
        int tmp = A->val;
        A->val = B->val;
        B->val = tmp;
    }
};

File: Difficulty/Medium/[92] Reverse Linked List II/solution.py
#
# @lc app=leetcode id=92 lang=python3
#
# [92] Reverse Linked List II
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if m == n:
            return head
        m, n, pre, count = min(m, n), max(m, n), [], 1
        new_head = ListNode(0)
        new_head.next = head
        while head is not None:
            if count >= m and count < (n + m + 1) // 2:
                pre.append(head)
            if count > (m + n) // 2 and count <= n:
                head.val, pre[n - count].val = pre[n - count].val, head.val
            count += 1
            head = head.next
        return new_head.next
        
        

